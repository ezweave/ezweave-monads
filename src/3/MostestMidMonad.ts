interface IMonad<A> {
  bind<B>(callback: (value?: A) => Monad<B>): IMonad<B>;
  map<B>(callback: (value?: A) => B): IMonad<B>;
}

abstract class Monad<A> implements IMonad<A> {
  abstract bind<B>(fn: (value?: A) => IMonad<B>): IMonad<B>;
  abstract map<B>(callback: (value?: A) => B): IMonad<B>;

  public static some<A>(value: A) {
    return new MostestMidMonad<A>(value);
  }

  public static none() {
    return new MostestMidMonad<undefined>();
  }
}

export class MostestMidMonad<A> extends Monad<A> {
  myValue?: A;

  constructor(t?: A) {
    super();
    if (t) {
      this.myValue = t;
    }
  }

  public bind<B>(callback: (value?: A) => IMonad<B>) {
    return callback(this.myValue);
  }

  public map<B>(callback: (value?: A) => B) {
    return MostestMidMonad.some(callback(this.myValue));
  }
}
