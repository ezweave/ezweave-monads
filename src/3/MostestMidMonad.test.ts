import { toUpper } from 'lodash';

import { MostestMidMonad } from './MostestMidMonad';

describe(MostestMidMonad, () => {
  it('can use map', () => {
    const two = MostestMidMonad.some('foo')
      .map(toUpper)
      .map((n) => `${n}_MOSTEST_MID`);

    expect(two).toMatchSnapshot();
  });
});
