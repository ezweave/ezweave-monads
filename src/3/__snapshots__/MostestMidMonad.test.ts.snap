// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MostestMidMonad can use map 1`] = `
MostestMidMonad {
  "value": "FOO_MOSTEST_MID",
}
`;
